### PR TITLE
feat: update leveragelp pool

### PR DESCRIPTION
**Pull request by cryptokage1996**
## Description

Closes:
*https://trello.com/c/HwkuOSkf/403-leveragelp-pool-update

### What has Changed?

updated tx for leveragelp disable or enable existing pool
added tx for dding new pools 

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/elys-network/elys/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/elys-network/elys/blob/main/docs/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/elys-network/elys/blob/main/CONTRIBUTING.md#testing)
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### Deployment Notes

Are there any specific considerations to take into account when deploying these changes? This may include new dependencies, scripts that need to be executed, or any aspects that can only be evaluated in a deployed environment.

### Screenshots and Videos

Please provide any relevant before and after screenshots by uploading them here. Additionally, demo videos can be highly beneficial in demonstrating the process.

### CODE REVIEW

File: x/leveragelp/client/cli/tx_add_pool.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+func readPoolJSON(filename string) (types.Pool, error) {
+	var pool types.Pool
+	bz, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return types.Pool{}, err
+	}
+	err = json.Unmarshal(bz, &pool)
+	if err != nil {
+		return types.Pool{}, err
```
Comment (amityadav0) [inline x/leveragelp/client/cli/tx_add_pool.go:105]: When creating new pool, I think we should use default values for lpAmount and health. In current case governance can set health or lpAmount to anything

File: x/leveragelp/keeper/msg_server_add_pool.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+)
+
+func (k msgServer) AddPools(goCtx context.Context, msg *types.MsgAddPool) (*types.MsgAddPoolResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	if k.authority != msg.Authority {
+		return nil, errorsmod.Wrapf(govtypes.ErrInvalidSigner, "invalid authority; expected %s, got %s", k.authority, msg.Authority)
+	}
+
+	_, found := k.GetPool(ctx, msg.Pool.AmmPoolId)
```
Comment (amityadav0) [inline x/leveragelp/keeper/msg_server_add_pool.go:19]: Please add a check here `Pool.AmmPoolId should be oracle enabled`

File: x/leveragelp/keeper/msg_server_remove_pool.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+func (k msgServer) RemovePool(goCtx context.Context, msg *types.MsgAddPool) (*types.MsgAddPoolResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	if k.authority != msg.Authority {
+		return nil, errorsmod.Wrapf(govtypes.ErrInvalidSigner, "invalid authority; expected %s, got %s", k.authority, msg.Authority)
+	}
+
+	_, found := k.GetPool(ctx, msg.Pool.AmmPoolId)
+
```
Comment (amityadav0) [inline x/leveragelp/keeper/msg_server_remove_pool.go:20]: Please add some checks here, we should not be able to delete Pool with lpAmount > 0

File: x/leveragelp/keeper/msg_server_remove_pool.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import (
+	"context"
+
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+	"github.com/elys-network/elys/x/leveragelp/types"
+)
+
+func (k msgServer) RemovePool(goCtx context.Context, msg *types.MsgAddPool) (*types.MsgAddPoolResponse, error) {
```
Comment (amityadav0) [inline x/leveragelp/keeper/msg_server_remove_pool.go:12]: Why are we using `types.MsgAddPool ` here ?

File: x/leveragelp/keeper/msg_server_update_pools.go [hunk 5]
```diff
@@ -16,7 +16,11 @@ func (k msgServer) UpdatePools(goCtx context.Context, msg *types.MsgUpdatePools)
 		return nil, errorsmod.Wrapf(govtypes.ErrInvalidSigner, "invalid authority; expected %s, got %s", k.authority, msg.Authority)
 	}
 
-	for _, pool := range msg.Pools {
+	pool, found := k.GetPool(ctx, msg.Pool.AmmPoolId)
+
+	if found {
+		pool.Enabled = msg.Pool.Enabled
+		pool.Closed = msg.Pool.Closed
```
Comment (amityadav0) [inline x/leveragelp/keeper/msg_server_update_pools.go:23]: I think we don't need whole pool object values when updating, please update it to use required values which are `Enabled` / `Closed` and `PoolId`

File: x/leveragelp/keeper/msg_server_update_pools.go [hunk 6]
```diff
@@ -16,7 +16,11 @@ func (k msgServer) UpdatePools(goCtx context.Context, msg *types.MsgUpdatePools)
 		return nil, errorsmod.Wrapf(govtypes.ErrInvalidSigner, "invalid authority; expected %s, got %s", k.authority, msg.Authority)
 	}
 
-	for _, pool := range msg.Pools {
+	pool, found := k.GetPool(ctx, msg.Pool.AmmPoolId)
```
Comment (amityadav0) [inline x/leveragelp/keeper/msg_server_update_pools.go:19]: This should be pool id of pool in leveragelp

Comment (amityadav0) [reply x/leveragelp/keeper/msg_server_update_pools.go:19]: not AmmPoolId in leveragelp

File: x/leveragelp/keeper/msg_server_remove_pool.go [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	if k.authority != msg.Authority {
+		return nil, errorsmod.Wrapf(govtypes.ErrInvalidSigner, "invalid authority; expected %s, got %s", k.authority, msg.Authority)
+	}
+
+	_, found := k.GetPool(ctx, msg.Pool.AmmPoolId)
+
+	if found {
+		k.DeletePool(ctx, msg.Pool.AmmPoolId)
```
Comment (amityadav0) [inline x/leveragelp/keeper/msg_server_remove_pool.go:22]: we should accept a single number which is pool id instead of whole pool object

**Review by amityadav0 — CHANGES_REQUESTED** [review]

File: x/leveragelp/keeper/msg_server_update_pools.go (hunk 6 above)
Comment (amityadav0) [reply x/leveragelp/keeper/msg_server_update_pools.go:19]: thanks

File: x/leveragelp/keeper/msg_server_update_pools.go (hunk 5 above)
Comment (amityadav0) [reply x/leveragelp/keeper/msg_server_update_pools.go:23]: okay, we don't need `pool` object here

File: x/leveragelp/keeper/msg_server_add_pool.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+	if found && pool.PoolParams.UseOracle {
+		_, found := k.GetPool(ctx, msg.Pool.AmmPoolId)
+
+		if !found {
+			newPool.AmmPoolId = msg.Pool.AmmPoolId
+			newPool.Closed = msg.Pool.Closed
+			newPool.Enabled = msg.Pool.Enabled
+			newPool.LeverageMax = msg.Pool.LeverageMax
+			k.SetPool(ctx, newPool)
```
Comment (amityadav0) [inline x/leveragelp/keeper/msg_server_add_pool.go:30]: need to initialise other values too, lpAmount should be zero etc

File: x/leveragelp/types/errors.go [hunk 9]
```diff
@@ -31,4 +31,5 @@ var (
 	ErrNegUserAmountAfterRepay = errorsmod.Register(ModuleName, 38, "negative user amount after repay")
 	ErrInvalidLeverage         = errorsmod.Register(ModuleName, 39, "leverage should be same as existing position")
 	ErrInvalidCollateral       = errorsmod.Register(ModuleName, 40, "collateral should not be more than total liability")
+	ErrPoolLevrageAmountNotZero= errorsmod.Register(ModuleName, 41, "pools leverage amount is grater than zero")
```
Comment (amityadav0) [inline x/leveragelp/types/errors.go:34]:
```suggestion
	ErrPoolLeverageAmountNotZero= errorsmod.Register(ModuleName, 41, "pool leverage amount is greater than zero")
```

**Comment by amityadav0** [comment]
Please add screenshots of these operations after testing.

**Review by cosmic-vagabond — APPROVED** [review]

**Pull request merged**